[TTI-16] Add rust message serialization

jbangelo opened the pull request:
The approach taken is to define a trait (`SbpSerialize`) which defines functions for how a type can be serialized according to the SBP specification. Several basic types have the serialize trait implemented, along with any type in the YAML spec which doesn't have a message ID associated with it.  I also added an `into_frame()` function to the  `SBPMessage` trait which consumes the message struct and converts it into a `Vec<u8>` representing a full SBP frame. The message definitions in the YAML spec have the `into_frame` function defined for them.

In addition this PR adds code to implement the  the `SBPMessage` trait for the `SBP` enum. This should make it easier to handle generic messages coming out of the parser, along with handling specific messages coming out of the parser using a `match` expression. It also changes the message type from an associated `const` value to be accessible via a function, this should make it possible to make trait objects for the `SBPMessage` trait (nothing in the PR requires it, but this limitation was noticed while testing).

benjaminaltieri commented on generator/sbpg/targets/resources/sbp_messages_mod.rs:
```diff
@@ -57,3 +58,55 @@ impl SBP {
+                crc.update(&frame[1..]);
+                frame.extend(crc.get().into_sbp_bytes());
+
+                frame
```
regardless of how we end up doing framing we need to check that the total size is less than the max payload size of SBP

jbangelo replied:
Done! I ended up added a whole separate error and result type just for the framing.

jbangelo commented on generator/sbpg/targets/resources/sbp_messages_template.rs:
```diff
@@ -89,6 +94,55 @@ impl super::SBPMessage for (((m.identifier|camel_case))) {
         self.sender_id = Some(new_id);
     }
+
+    fn into_frame(self) -> Vec<u8> {
```
We could define a default implementation of `into_frame` method in the `SBPMessage` trait so we don't have to re-implement it in each message. But you're right that we'd still have to implement `SBPSerialize` for each message anyways.

jbangelo replied:
Never mind, it seems as though you can't give a default implementation of a "consuming" method of a trait, since the size of the `self` isn't known in the general case :(

benjaminaltieri replied:
Ah ok yeah I knew there was probably as catch. After writing these initial comments I was thinking the framing could just be a regular old static method that has those trait requirements which the concrete implementation would call, but there is little benefits at that point other than code organization since ultimately those are generated sources.

jbangelo commented:
I've pushed an updated approach to things.

- The `SBP` enum now has a function to get a `&dyn SBPMessage` reference to the underlying message. This was the best I could figure out how to do this, and it still ended up with a giant match expression. @silverjam I would love to talk with you next week about alternative approaches, I went through a bunch and ended up with this.
- I've added a dedicated `Unknown` message struct for holding unknown message types.
- I've reorganized the serialize trait to put into a vector instead of returning a new vector. This makes it a little bit more clunky to use but I think will make the code more effecicent.
 - I've added a `framer` module to contain the logic of creating a `Vec<u8>` representing a frame from a message object. The framing function no longer consumes the message, this is because it takes in a `&dyn SBPMessage` to make it easier to use with the `SBP` enum.

jbangelo commented on generator/sbpg/targets/resources/sbp_messages_mod.rs:
```diff
@@ -22,21 +22,22 @@ use self::(((p.identifier|mod_name)))::(((m.identifier|camel_case)));
+    fn get_message_type(&self) -> u16;
     fn get_sender_id(&self) -> Option<u16>;
     fn set_sender_id(&mut self, new_id: u16);
+    fn into_frame(self) -> Vec<u8>;
```
I ended up just taking a `&dyn SBPMessage` to the framing function and making the `SBPMessage` inherit/extend the `SbpSerialize` trait.

benjaminaltieri replied:
This is great.

Just for fun I played around a bit to see if we could bring back the `SBPMessage::to_frame` trait method for ergonomics' sake. It's basically the same as before in terms of code duplication since it can't be a default trait implementation due to the `Sized` limitation you mentioned before, but at least it's effectively the same implementation with dispatches to `sbp::framer::to_frame` instead of having different generated implementations. Curious what you guys think.

https://github.com/swift-nav/libsbp/pull/758
